File: src/clients/desktop_app/ami_daemon/base_agent/tools/eigent_browser/page_snapshot.py
"""
PageSnapshot - Utility for capturing YAML-like page snapshots.

Ported from CAMEL-AI/Eigent project.
Converts DOM into LLM-friendly text format with element references [ref=eN].
"""

import logging
from pathlib import Path
from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union

if TYPE_CHECKING:
    from playwright.async_api import Page

from .config_loader import ConfigLoader

logger = logging.getLogger(__name__)


class PageSnapshot:
    """Utility for capturing YAML-like page snapshots and diff-only variants."""

    def __init__(self, page: "Page"):
        self.page = page
        self.snapshot_data: Optional[str] = None  # last full snapshot
        self._last_url: Optional[str] = None
        self.last_info: Dict[str, List[int] | bool] = {
            "is_diff": False,
            "priorities": [1, 2, 3],
        }
        self.dom_timeout = ConfigLoader.get_dom_content_loaded_timeout()

    # ---------------------------------------------------------------------
    # Public API
    # ---------------------------------------------------------------------
    async def capture(
        self,
        *,
        force_refresh: bool = False,
        diff_only: bool = False,
        viewport_limit: bool = False,
    ) -> str:
        """Return current snapshot or just the diff to previous one."""
        try:
            current_url = self.page.url

            # ensure DOM stability
            await self.page.wait_for_load_state(
                'domcontentloaded', timeout=self.dom_timeout
            )

            logger.debug("Capturing page snapshot...")
            snapshot_result = await self._get_snapshot_direct(
                viewport_limit=viewport_limit
            )

            # Extract snapshot text from the unified analyzer result
            if (
                isinstance(snapshot_result, dict)
                and 'snapshotText' in snapshot_result
            ):
                snapshot_text = snapshot_result['snapshotText']
            else:
                snapshot_text = snapshot_result

            formatted = self._format_snapshot(snapshot_text or "<empty>")

            output = formatted
            if diff_only and self.snapshot_data:
                output = self._compute_diff(self.snapshot_data, formatted)

            # update cache with *full* snapshot (not diff)
            self._last_url = current_url
            self.snapshot_data = formatted

            # analyse priorities present (only for non-diff)
            priorities_included = self._detect_priorities(
                formatted if not diff_only else self.snapshot_data or formatted
            )
            self.last_info = {
                "is_diff": diff_only and self.snapshot_data is not None,
                "priorities": priorities_included,
            }

            logger.debug(
                "Snapshot captured. Diff_only=%s, priorities=%s",
                diff_only,
                self.last_info["priorities"],
            )
            return output
        except Exception as exc:
            logger.error("Snapshot capture failed: %s", exc)
            return f"Error: Could not capture page snapshot {exc}"

    async def get_full_result(
        self,
        *,
        viewport_limit: bool = False,
    ) -> Dict[str, Any]:
        """Return the full analysis result including elements map."""
        try:
            await self.page.wait_for_load_state(
                'domcontentloaded', timeout=self.dom_timeout
            )

            result = await self._get_snapshot_direct(viewport_limit=viewport_limit)
            if isinstance(result, dict):
                return result
            return {"snapshotText": result, "elements": {}}
        except Exception as exc:
            logger.error("Full snapshot capture failed: %s", exc)
            return {"snapshotText": f"Error: {exc}", "elements": {}}

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------
    _snapshot_js_cache: Optional[str] = None  # class-level cache

    async def _get_snapshot_direct(
        self, viewport_limit: bool = False
    ) -> Optional[Union[str, Dict[str, Any]]]:
        """Evaluate the snapshot-extraction JS with simple retry logic."""

        # Load JS once and cache it at class level
        if PageSnapshot._snapshot_js_cache is None:
            js_path = Path(__file__).parent / "unified_analyzer.js"
            PageSnapshot._snapshot_js_cache = js_path.read_text(encoding="utf-8")

        js_code = PageSnapshot._snapshot_js_cache

        retries: int = 3
        while retries > 0:
            try:
                return await self.page.evaluate(js_code, viewport_limit)
            except Exception as e:
                msg = str(e)

                # Typical error when navigation happens between calls
                nav_err = "Execution context was destroyed"

                if (
                    nav_err in msg
                    or "Most likely because of a navigation" in msg
                ):
                    retries -= 1
                    logger.debug(
                        "Snapshot evaluate failed due to navigation; "
                        "retrying (%d left)...",
                        retries,
                    )

                    # Wait for next DOM stability before retrying
                    try:
                        await self.page.wait_for_load_state(
                            "domcontentloaded", timeout=self.dom_timeout
                        )
                    except Exception:
                        # Even if waiting fails, attempt retry
                        pass

                    continue  # retry the evaluate()

                # Any other exception -> abort
                logger.warning(
                    "Failed to execute snapshot JavaScript: %s",
                    e,
                )
                return None

        logger.warning("Failed to execute snapshot JavaScript after retries")
        return None

    @staticmethod
    def _format_snapshot(text: str) -> str:
        return "\n".join(["- Page Snapshot", "```yaml", text, "```"])

    @staticmethod
    def _compute_diff(old: str, new: str) -> str:
        if not old or not new:
            return "- Page Snapshot (error: missing data for diff)"

        import difflib

        diff = list(
            difflib.unified_diff(
                old.splitlines(False),
                new.splitlines(False),
                fromfile='prev',
                tofile='curr',
                lineterm='',
            )
        )
        if not diff:
            return "- Page Snapshot (no structural changes)"
        return "\n".join(["- Page Snapshot (diff)", "```diff", *diff, "```"])

    # ------------------------------------------------------------------
    def _detect_priorities(self, snapshot_yaml: str) -> List[int]:
        """Return sorted list of priorities present (1,2,3)."""
        priorities = set()
        for line in snapshot_yaml.splitlines():
            if '[ref=' not in line:
                continue
            lower_line = line.lower()
            if any(
                r in lower_line
                for r in (
                    "input",
                    "button",
                    "select",
                    "textarea",
                    "checkbox",
                    "radio",
                    "link",
                )
            ):
                priorities.add(1)
            elif 'label' in lower_line:
                priorities.add(2)
            else:
                priorities.add(3)
        if not priorities:
            priorities.add(3)
        return sorted(priorities)

File: src/clients/desktop_app/daemon-ts/src/browser/page-snapshot.ts
/**
 * PageSnapshot — Captures YAML-like page snapshots using unified_analyzer.js.
 *
 * Ported from page_snapshot.py.
 *
 * Key concepts:
 * - Evaluates unified_analyzer.js in page context
 * - Returns YAML-like accessibility tree with [ref=eN] element references
 * - Diff support for incremental updates
 * - Retry for navigation-destroyed contexts
 */

import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import type { Page } from "playwright";
import { BrowserConfig } from "./config.js";
import { createLogger } from "../utils/logging.js";

const logger = createLogger("page-snapshot");

// ===== JS cache (module-level singleton) =====

let _snapshotJsCache: string | null = null;

function getSnapshotJs(): string {
  if (_snapshotJsCache !== null) return _snapshotJsCache;

  // Resolve path relative to this file
  const __filename = fileURLToPath(import.meta.url);
  const __dirname = dirname(__filename);

  // unified_analyzer.js lives alongside this file in scripts/
  const candidates = [
    resolve(__dirname, "scripts/unified_analyzer.js"),
    // Fallback: relative to daemon-ts/dist/browser/ at runtime
    resolve(__dirname, "../src/browser/scripts/unified_analyzer.js"),
  ];

  for (const candidate of candidates) {
    try {
      _snapshotJsCache = readFileSync(candidate, "utf-8");
      logger.info({ path: candidate }, "Loaded unified_analyzer.js");
      return _snapshotJsCache;
    } catch {
      // try next
    }
  }

  throw new Error(
    `Could not find unified_analyzer.js. Tried: ${candidates.join(", ")}`,
  );
}

// ===== Snapshot result types =====

interface SnapshotResult {
  snapshotText: string;
  elements: Record<string, unknown>;
  metadata?: {
    refDebug?: {
      weakmapHit?: number;
      ariaRefHit?: number;
      signatureHit?: number;
      newRef?: number;
      evicted?: number;
    };
    refCounterValue?: number;
    totalMappedRefs?: number;
  };
}

// ===== PageSnapshot class =====

export class PageSnapshot {
  private page: Page;
  private snapshotData: string | null = null;
  private _lastUrl: string | null = null;
  lastInfo: { isDiff: boolean; priorities: number[] } = {
    isDiff: false,
    priorities: [1, 2, 3],
  };

  constructor(page: Page) {
    this.page = page;
  }

  // ===== Public API =====

  async capture(options?: {
    forceRefresh?: boolean;
    diffOnly?: boolean;
    viewportLimit?: boolean;
  }): Promise<string> {
    const diffOnly = options?.diffOnly ?? false;
    const viewportLimit = options?.viewportLimit ?? false;

    try {
      const currentUrl = this.page.url();

      // Ensure DOM stability
      await this.page.waitForLoadState("domcontentloaded", {
        timeout: BrowserConfig.domLoadedTimeout,
      });

      logger.debug("Capturing page snapshot...");
      const snapshotResult = await this._getSnapshotDirect(viewportLimit);

      // Extract snapshot text
      let snapshotText: string;
      if (
        snapshotResult &&
        typeof snapshotResult === "object" &&
        "snapshotText" in snapshotResult
      ) {
        const result = snapshotResult as SnapshotResult;
        snapshotText = result.snapshotText;

        // Log ref debug info
        const metadata = result.metadata;
        if (metadata?.refDebug) {
          const rd = metadata.refDebug;
          logger.debug(
            {
              weakmapHit: rd.weakmapHit,
              ariaRefHit: rd.ariaRefHit,
              signatureHit: rd.signatureHit,
              newRef: rd.newRef,
              evicted: rd.evicted,
              refCounter: metadata.refCounterValue,
              totalMapped: metadata.totalMappedRefs,
            },
            "Ref assignment stats",
          );
        }
      } else {
        snapshotText = snapshotResult as string;
      }

      const formatted = PageSnapshot._formatSnapshot(snapshotText || "<empty>");

      let output = formatted;
      if (diffOnly && this.snapshotData) {
        output = PageSnapshot._computeDiff(this.snapshotData, formatted);
      }

      // Update cache with full snapshot (not diff)
      this._lastUrl = currentUrl;
      this.snapshotData = formatted;

      // Analyse priorities
      const prioritiesIncluded = PageSnapshot._detectPriorities(
        diffOnly ? this.snapshotData || formatted : formatted,
      );
      this.lastInfo = {
        isDiff: diffOnly && this.snapshotData !== null,
        priorities: prioritiesIncluded,
      };

      logger.debug(
        { diffOnly, priorities: this.lastInfo.priorities },
        "Snapshot captured",
      );
      return output;
    } catch (exc) {
      logger.error({ err: exc }, "Snapshot capture failed");
      return `Error: Could not capture page snapshot ${exc}`;
    }
  }

  async getFullResult(options?: {
    viewportLimit?: boolean;
  }): Promise<Record<string, unknown>> {
    try {
      await this.page.waitForLoadState("domcontentloaded", {
        timeout: BrowserConfig.domLoadedTimeout,
      });

      const result = await this._getSnapshotDirect(options?.viewportLimit ?? false);
      if (result && typeof result === "object" && "snapshotText" in result) {
        return result as unknown as Record<string, unknown>;
      }
      return { snapshotText: result, elements: {} };
    } catch (exc) {
      logger.error({ err: exc }, "Full snapshot capture failed");
      return { snapshotText: `Error: ${exc}`, elements: {} };
    }
  }

  // ===== Internal helpers =====

  private async _getSnapshotDirect(
    viewportLimit: boolean,
  ): Promise<string | SnapshotResult | null> {
    const jsCode = getSnapshotJs();
    let retries = 3;

    while (retries > 0) {
      try {
        return await this.page.evaluate(jsCode, viewportLimit);
      } catch (e: unknown) {
        const msg = e instanceof Error ? e.message : String(e);
        const navErr = "Execution context was destroyed";

        if (
          msg.includes(navErr) ||
          msg.includes("Most likely because of a navigation")
        ) {
          retries--;
          logger.debug(
            { retriesLeft: retries },
            "Snapshot evaluate failed due to navigation; retrying",
          );

          try {
            await this.page.waitForLoadState("domcontentloaded", {
              timeout: BrowserConfig.domLoadedTimeout,
            });
          } catch {
            // Even if waiting fails, attempt retry
          }
          continue;
        }

        // Any other exception → abort
        logger.warn({ err: e }, "Failed to execute snapshot JavaScript");
        return null;
      }
    }

    logger.warn("Failed to execute snapshot JavaScript after retries");
    return null;
  }

  private static _formatSnapshot(text: string): string {
    return ["- Page Snapshot", "```yaml", text, "```"].join("\n");
  }

  private static _computeDiff(old: string, newStr: string): string {
    if (!old || !newStr) {
      return "- Page Snapshot (error: missing data for diff)";
    }

    const oldLines = old.split("\n");
    const newLines = newStr.split("\n");

    // Simple line-based diff
    const diffLines: string[] = [];
    const maxLen = Math.max(oldLines.length, newLines.length);

    for (let i = 0; i < maxLen; i++) {
      const oldLine = i < oldLines.length ? oldLines[i] : undefined;
      const newLine = i < newLines.length ? newLines[i] : undefined;

      if (oldLine === newLine) {
        // unchanged — skip for brevity
        continue;
      }
      if (oldLine !== undefined && newLine === undefined) {
        diffLines.push(`- ${oldLine}`);
      } else if (oldLine === undefined && newLine !== undefined) {
        diffLines.push(`+ ${newLine}`);
      } else if (oldLine !== newLine) {
        diffLines.push(`- ${oldLine}`);
        diffLines.push(`+ ${newLine}`);
      }
    }

    if (diffLines.length === 0) {
      return "- Page Snapshot (no structural changes)";
    }

    return ["- Page Snapshot (diff)", "```diff", ...diffLines, "```"].join("\n");
  }

  private static _detectPriorities(snapshotYaml: string): number[] {
    const priorities = new Set<number>();

    for (const line of snapshotYaml.split("\n")) {
      if (!line.includes("[ref=")) continue;
      const lowerLine = line.toLowerCase();

      if (
        ["input", "button", "select", "textarea", "checkbox", "radio", "link"].some(
          (r) => lowerLine.includes(r),
        )
      ) {
        priorities.add(1);
      } else if (lowerLine.includes("label")) {
        priorities.add(2);
      } else {
        priorities.add(3);
      }
    }

    if (priorities.size === 0) {
      priorities.add(3);
    }

    return [...priorities].sort();
  }
}
